### ISSUE TITLE
MoveType random is not frame accurate

### ISSUE THREAD

**Issue by fmatthew5876** [EasyRPG/Player#1638]
Some details in test case 10.3 in #1567 

move type random has some unique behaviors. Needs more investigation and fix to replicate RPG_RT

**Comment by fmatthew5876** [comment]
Test Cases
--------------

Setup a map event with `move_frequency=8`, `move_speed=6`, `move_type=random`

Make it trigger parallel and put this code:
```
Loop:
 OpenSaveMenu
```

Despite having `move_frequency==8`, we see sometimes the event waits for some stop count.
Here are some results for the number of frames waited between each step.

| Wait | Num Samples |
| 0 | 26 |
| 1 | 4 |

**Comment by CherryDT** [comment]
There is a table that maps frequency to `max_stop_count`:

|`move_frequency`|`max_stop_count`|
|-|-|
|1|256|
|2|128|
|3|64|
|4|32|
|5|16|
|6|8|
|7|4|
|8|0|

Now it seems that the `max_stop_count` that is set after each move is further modified with a multiplier based on the `move_type`:

|`move_type`|`max_stop_count` multiplier|
|-|-|
|`random`|(random(0...4) + 3) / 5|
|`custom`|1 or 0.5*|
|All other types|1|

*: This is a strange case which I discovered only now. From how it looks to me, the normal value is used after each move step, but **when a new event page is chosen in an event**, that event's `max_stop_count` is **once** set to half the normal value when the `move_type` is `custom`.

Then there is a probability table for the different move actions in case of `random` movement:

|Probability|Action|
|-|-|
|30%|Move forward|
|20%|Turn left, move forward|
|20%|Turn right, move forward|
|10%|Turn around, move forward|
|20%|Wait for random(0...`max_stop_count`)|

**Comment by fmatthew5876** [comment]
> *: This is a strange case which I discovered only now. From how it looks to me, the normal value is used after each move step, but **when a new event page is chosen in an event**, that event's `max_stop_count` is **once** set to half the normal value when the `move_type` is `custom`.

I discovered this in #1601. There is the `max_stop_count` table for stepping and the `max_stop_count` table for turning. For some reason `RPG_RT` defaults to the turning `max_stop_count` when the move type is custom. This is only temporary though. As soon as you do any movement action like taking a step, `max_stop_count` is reset to its proper table value for step, turn, or wait as needed.


> `random` 	(random(0...4) + 3) / 5
> 20% 	Wait for random(0...`max_stop_count`)

Are you sure about these? From my testing I observed something different.

With move type random, I saw the `max_stop_count` was always set to the stepping value from your table above. However I did notice that rarely (Maybe 20% of the time?) `stop_count` would tick up to `max_stop_count+1` and then a move action would trigger.

**Comment by CherryDT** [comment]
At least that's RM2k behavior, pretty sure. Not sure about RM2k3 though.

It's possible though that the multiplier for `random` doesn't even take effect because something else sets the normal value afterwards...

**Comment by fmatthew5876** [comment]
What about random step in a move route? Is it just direction = rand(0,3) or does it have a probability table like above?

**Comment by CherryDT** [comment]
It is simply a random turn with equal distribution plus a step forward.

**Comment by CherryDT** [comment]
Also I fixed a mistake above, it should be:

> Set `stop_count` to random(0...`max_stop_count`) (effectively waiting between 1 and `max_stop_count` (inclusive) frames)

instead of

> Wait for random(0...`max_stop_count`)

**Comment by fmatthew5876** [comment]
> It's possible though that the multiplier for `random` doesn't even take effect because something else sets the normal value afterwards...

That seems plausible. Just setting `max_stop_count` wouldn't do much as you'd also have to set `stop_count` back to 0.

Also in one of my tests I remember observing that during a step, the `max_stop_count` gets reset to the stepping value every frame.

All of that combined would lead to what I observed where it only waits 1 frame.

**Comment by CherryDT** [comment]
> you'd also have to set `stop_count` back to 0.
This happens automatically after each move action, I was only concerned about the value of `max_stop_count` now.

**Comment by Ghabry** [comment]
test

**Comment by fmatthew5876** [comment]
One more question, for determining direction here. Are we using `direction` or `sprite_direction`?

**Comment by CherryDT** [comment]
In what context?

**Comment by fmatthew5876** [comment]
In this table, the turn actions. What is the input direction used?

> Probability 	Action
> 30% 	Move forward
> 20% 	Turn left, move forward
> 20% 	Turn right, move forward
> 10% 	Turn around, move forward
> 20% 	Set `stop_count` to random(0...`max_stop_count`) (effectively waiting between 1 and `max_stop_count` (inclusive) frames)

**Comment by CherryDT** [comment]
The turn action in this table simply reads and writes `direction`, in the same way the regular turn action does (i.e. if it was a h/v direction then rotate, if it was a diagonal direction reset to `down`).
The move action in this table updates `sprite_direction` based on `direction` (unless `lock_facing` is set) with the usual rules (which basically means `sprite_direction` is set to `direction` if it is h/v, or based on a complex ruleset including the previous `sprite_direction` in case `direction` is diagonal, although the latter is irrelevant in our case).

**Comment by CherryDT** [comment]
Hm, even though it's not relevant here, maybe it's relevant somewhere else - this is the ruleset for updating the `sprite_direction` upon movement:

|`direction`|Previous `sprite_direction`|New `sprite_direction`|
|-|-|-|
|`Up`|irrelevant|`Up`|
|`Right`|irrelevant|`Right`|
|`Down`|irrelevant|`Down`|
|`Left`|irrelevant|`Left`|
|`UpRight`|`Up`|`Up`|
|`UpRight`|`Right`|`Right`|
|`UpRight`|`Down`|`Up`|
|`UpRight`|`Left`|`Right`|
|`DownRight`|`Up`|`Down`|
|`DownRight`|`Right`|`Right`|
|`DownRight`|`Down`|`Down`|
|`DownRight`|`Left`|`Right`|
|`DownLeft`|`Up`|`Down`|
|`DownLeft`|`Right`|`Left`|
|`DownLeft`|`Down`|`Down`|
|`DownLeft`|`Left`|`Left`|
|`UpLeft`|`Up`|`Up`|
|`UpLeft`|`Right`|`Left`|
|`UpLeft`|`Down`|`Up`|
|`UpLeft`|`Left`|`Left`|

**Comment by fmatthew5876** [comment]
So coming back to this now armed with some dynrpg dumps. I made a map with 8 random moving events, each with frequency 1 to 8. Every frame I logged their data and let it go for a while.

There is something funky going on here.

Out of 766704 samples, 2120 of them have `stop_count > max_stop_count` which should be impossible.

In particular for the event with `frequency == 8`, `max_stop_count = 0` always which is expected. But for a few samples I see `stop_count` higher. In one case, `max_stop_count == 0` and `stop_count == 5`

I've uploaded my dataset here for posterity. It's using dynrpg names of `stepFrameCounter` and `frequencyTimer`. The event `id` is equal to it's movement frequency.

[chardump-large.txt](https://github.com/EasyRPG/Player/files/4459249/chardump-large.txt)


More investigation to come..
